resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesClientBuilder.scala

```diff
@@ -78,6 +81,25 @@ private[spark] class KubernetesClientBuilder(sparkConf: SparkConf, namespace: St
       }
       serviceAccountConfigBuilder
     }
-    new DefaultKubernetesClient(configBuilder.build)
+    val threadPoolExecutor = new Dispatcher().executorService().asInstanceOf[ThreadPoolExecutor]
+    // Set threads to be daemons in order to allow the driver main thread
+    // to shut down upon errors. Otherwise the driver will hang indefinitely.
+    threadPoolExecutor.setThreadFactory(new ThreadFactory {
+        override def newThread(r: Runnable): Thread = {
+          val thread = new Thread(r, "spark-on-k8s")
+          thread.setDaemon(true)
+          thread
+        }
+      })
+    // Disable the ping thread that is not daemon, in order to allow
+    // the driver main thread to shut down upon errors. Otherwise, the driver
+    // will hang indefinitely.
+    val config = configBuilder
+      .withWebsocketPingInterval(0)
```

Comment:
@ash211 @foxish 
I looked at the full stack traces, but they don't tell you much as to who submitted the threads. (Because submission takes place in other threads)

For the web socket ping interval, I believe this is the relevant source code. From [RealWebSocket.java](https://github.com/square/okhttp/blob/e172706b56a8616cf70f8d8285d5f8701c8a36a0/okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.java):
```
  public void initReaderAndWriter(
      String name, long pingIntervalMillis, Streams streams) throws IOException {
    synchronized (this) {
      this.streams = streams;
      this.writer = new WebSocketWriter(streams.client, streams.sink, random);
      this.executor = new ScheduledThreadPoolExecutor(1, Util.threadFactory(name, false));
      if (pingIntervalMillis != 0) {
        executor.scheduleAtFixedRate(
            new PingRunnable(), pingIntervalMillis, pingIntervalMillis, MILLISECONDS);
      }
      if (!messageAndCloseQueue.isEmpty()) {
        runWriter(); // Send messages that were enqueued before we were connected.
      }
    }
```
Notice it is calling `Util.threadFactory(name, false)`. The second parameter `false` sets a new thread to be non-daemon. So passing `true` will fix the issue.

For the other one, I think this is the source code. From [Dispatcher.java](https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/Dispatcher.java):
```
  public synchronized ExecutorService executorService() {
    if (executorService == null) {
      executorService = new ThreadPoolExecutor(0, Integer.MAX_VALUE, 60, TimeUnit.SECONDS,
          new SynchronousQueue<Runnable>(), Util.threadFactory("OkHttp Dispatcher", false));
    }
    return executorService;
  }
```

Again, passing `true` to `Util.threadFactory` will fix the issue.

Unfortunately, both are inside the `okhttp` project outside the k8s client code. Interestingly, the comment of the main class [OkHttpClient](https://github.com/square/okhttp/blob/36a4d16b3d7b47ca83624a3bff66cde0cb5fe9f5/okhttp/src/main/java/okhttp3/OkHttpClient.java) claims it is using daemon threads. So the above two lines could be simply bugs:

>  <p>OkHttp also uses daemon threads for HTTP/2 connections. These will exit automatically if they remain idle.